engines/lab/anim.cpp

```diff
@@ -0,0 +1,340 @@
+				bool didTOF = false;
+
+				if (_waitForEffect) {
+					while (_vm->_music->isSoundEffectActive()) {
+						_vm->_music->updateMusic();
+						_vm->waitTOF();
+
+						if (_vm->_graphics->_dispBitMap->_drawOnScreen)
+							didTOF = true;
+					}
+				}
+
+				_isPlaying = false;
+				_vm->_event->mouseShow();
+
+				if (!didTOF)
+					_vm->_graphics->screenUpdate();
+
+				return;
+			}
+
+			// Random frame number so it never gets back to 2
+			_frameNum = 4;
+			_diffFile->seek(_diffFileStart, SEEK_SET);
+			break;
+
+		default:
+			_diffFile->skip(_size);
+			break;
+		}
+	}
+}
+
+void Anim::stopDiff() {
+	if (_isPlaying && _isAnim)
+		_vm->_graphics->blackScreen();
+}
+
+void Anim::stopDiffEnd() {
+	if (!_isPlaying)
+		return;
+
+	_stopPlayingEnd = true;
+	while (_isPlaying) {
+		_vm->_music->updateMusic();
+		diffNextFrame();
+	}
+}
+
+void Anim::readDiff(Common::File *diffFile, bool playOnce, bool onlyDiffData) {
+	_playOnce = playOnce;
+	_delayMicros = 0;
+	_header = 0;
+	_curBit = 0;
+	_frameNum = 0;
+	_numChunks = 1;
+	_donePal = false;
+	_stopPlayingEnd = false;
+	_isPlaying = true;
+
+	if (_doBlack) {
+		_doBlack = false;
+		_vm->_graphics->blackScreen();
+	}
+
+	_diffFile = diffFile;
+
+	_continuous = false;
+	uint32 signature = _diffFile->readUint32BE();
+	_header = _diffFile->readUint32LE();
+
+	if ((signature != MKTAG('D', 'I', 'F', 'F')) || (_header != 1219009121L)) {
+		_isPlaying = false;
+		return;
+	}
+
+	_header = _diffFile->readUint32LE();
+	_size = _diffFile->readUint32LE();
+
+	if (_header == 0) {
+		// sizeof(headerdata) != 18, but the padding might be at the end
+		// 2 bytes, version, unused.
+		_diffFile->skip(2);
+		_headerdata._width = _diffFile->readUint16LE();
+		_headerdata._height = _diffFile->readUint16LE();
+		// 1 byte, depth, unused
+		_diffFile->skip(1);
+		_headerdata._fps = _diffFile->readByte();
+
+		// HACK: The original game defines a 1 second delay when changing screens, which is
+		// very annoying. We first removed the delay, but it looked wrong when changing screens
+		// as it was possible to see that something was displayed, without being able to tell
+		// what it was. A shorter delay (150ms) makes it acceptable during gameplay and
+		// readable. The big question is: do we need that message?
+		g_system->delayMillis(150);
+
+		if (_headerdata._fps == 1)
+			_headerdata._fps = 0;
+
+		// 4 + 2 bytes, buffer size and machine, unused
+		_diffFile->skip(6);
+		_headerdata._flags = _diffFile->readUint32LE();
+
+		_diffFile->skip(_size - 18);
+
+		_continuous = CONTINUOUS & _headerdata._flags;
+		_diffWidth = _headerdata._width;
+		_diffHeight = _headerdata._height;
+		_vm->_utils->setBytesPerRow(_diffWidth);
+
+		_numChunks = (((int32)_diffWidth) * _diffHeight) / 0x10000;
+
+		if ((uint32)(_numChunks * 0x10000) < (uint32)(((int32)_diffWidth) * _diffHeight))
+			_numChunks++;
+	} else
+		return;
+
+	for (_header = 0; _header < 8; _header++)
```

Comment:
Not sure why `this->_header` is used as an index here. Is it to conveniently reset `_header` for the use a few lines below?